poc/fiat_shamir.sage

```diff
@@ -45,48 +45,60 @@ class NIZK:
 
     def prove(self, witness, rng):
         """
-        Proving method using commitment-response format.
-
-        Allows for batching.
+        Proving method using challenge-response format.
         """
         (commitment, challenge, response) = self._prove(witness, rng)
-        # running the verifier here is just a sanity check
         assert self.sigma_protocol.verifier(commitment, challenge, response)
-        return self.sigma_protocol.serialize_commitment(commitment) + self.sigma_protocol.serialize_response(response)
-
+        domain_sep = "NISigmaProtocolShortProof".encode('utf-8')
+        return domain_sep + self.sigma_protocol.serialize_challenge(challenge) + self.sigma_protocol.serialize_response(response)
 
     def verify(self, proof):
-        # Before running the sigma protocol verifier, one must also check that:
-        # - the proof length is exactly commit_bytes_len + response_bytes_len
-        assert len(proof) == self.sigma_protocol.instance.commit_bytes_len + self.sigma_protocol.instance.response_bytes_len, f"Invalid proof length: {len(proof)} != {self.sigma_protocol.instance.commit_bytes_len + self.sigma_protocol.instance.response_bytes_len}"
-        # - proof deserialization successfully produces a valid commitment and a valid response
-        commitment_bytes = proof[:self.sigma_protocol.instance.commit_bytes_len]
-        response_bytes = proof[self.sigma_protocol.instance.commit_bytes_len:]
-        commitment = self.sigma_protocol.deserialize_commitment(commitment_bytes)
+        """
+        Verification method using challenge-response format.
+        """
+        # Check that serialized proof domain separator matches expected value
+        domain_sep = "NISigmaProtocolShortProof".encode('utf-8')
+        domain_sep_bytes = proof[:len(domain_sep)]
+        assert domain_sep_bytes == domain_sep
+
+        challenge_len = self.sigma_protocol.instance.Domain.scalar_byte_length()
+        challenge_bytes = proof[len(domain_sep):challenge_len]
+        response_bytes = proof[len(domain_sep) + challenge_len:]
+
+        challenge = self.sigma_protocol.deserialize_challenge(challenge_bytes)
         response = self.sigma_protocol.deserialize_response(response_bytes)
+        commitment = self.sigma_protocol.simulate_commitment(response, challenge)
 
-        self.codec.prover_message(self.hash_state, commitment)
-        challenge = self.codec.verifier_challenge(self.hash_state)
         return self.sigma_protocol.verifier(commitment, challenge, response)
 
-    def prove_short(self, witness, rng):
+    def prove_batchable(self, witness, rng):
         """
-        Alternative proving method using challenge-response format.
+        Proving method using commitment-response format.
+
+        Allows for batching.
         """
         (commitment, challenge, response) = self._prove(witness, rng)
+        # running the verifier here is just a sanity check
```

Comment:
nit: This PR deletes this comment above, but adds it here. Perhaps keep the comment in both for consistency?

Reply:
Bump